### PR TITLE
MRG, ENH: Support path-like objects

### LINKED ISSUE #6699: Further functions & methods that currently do not work with pathlib.Path objects
#6682 has already improved support for `str`-like paths,  like `pathlib.Path` objects. I just ran my source analysis stack again and found many more places where currently only strings are supported, and passing `pathlib.Path` objects instead raises  an error. I'm not sure if this will be really helpful to anyone, but I thought sharing the results with you at least wouldn't hurt :)

#### Plotting
- `mne.viz.plot_alignment()`: `trans` param

#### Source Spaces
- `mne.write_source_spaces()`: `fname` param
- `mne.read_source_spaces()`: `fname` param

#### BEM
- `mne.write_bem_surfaces()`: `fname` param
- `mne.write_bem_solution()`: `fname` param

#### Forward Solution
- `mne.make_forward_solution()`: `src` and `trans` params
- `mne.write_forward_solution()`: `fname` param

#### Noise Covariance Matrix
- `mne.write_cov()`: `fname` param
- `mne.read_cov():`: `fname` param

#### Inverse Operator
- `mne.minimum_norm.write_inverse_operator()`: `fname` param
- `mne.minimum_norm.read_inverse_operator()`: `fname` param

#### Source Estimate
- `mne.SourceEstimate.save()`: `fname` param
- `mne.read_source_estimate()`: `fname` param

### PR DESCRIPTION
**Pull request by larsoner**
Closes #6699.

- Unifies `preload`-as-`str` handling in `io/base.py`
- Adds `path-like` to `_validate_types`
- Makes `_validate_types` support nested `str` options (e.g. `(None, 'path-like')`)
- Modernizes a number of `tmpdir` uses (@massich you should be happy :) )

### CODE REVIEW

**Comment by massich** [comment]
+1000 :)

**Comment by agramfort** [comment]
you have a legit error https://travis-ci.org/mne-tools/mne-python/jobs/575530698

**Comment by larsoner** [comment]
Oh so sad on 3.5:
```
    def isfile(path):
        """Test whether a path is a regular file"""
        try:
>           st = os.stat(path)
E           TypeError: argument should be string, bytes or integer, not LocalPath
```
I don't see a way around this other than reverting all the changes I made to the tests to modernize them :(

**Comment by agramfort** [comment]
can you help me understand the issue?

**Comment by larsoner** [comment]
`os` does not understand `py._path.*` objects before 3.6, probably something about how they check for `str`-like-ness. Even `open(fname)` of these objects does not work on 3.5.

I realized I can work around it by monkey patching `os` in our `conftest.py`, though. So I think in the end we'll be okay.

**Comment by agramfort** [comment]
you tell me option 1 is put back the str and option 2 is monkey patching os
module from std lib?
If so please put back the str.

>

**Comment by larsoner** [comment]
> option 2 is monkey patching os module from std lib?

Just during tests, it's actually not too bad. All it does is wrap these classes to `str` if they are detected. See latest commit

**Comment by larsoner** [comment]
(Basically it allows us to write nicer, future-compatible code while working around a Python3.5 limitation. I don't see any real risk caused by the fixture I wrote.)

File: mne/forward/tests/test_forward.py [hunk 1]
```diff
@@ -377,7 +373,7 @@ def test_average_forward_solution():
 
     # modify a fwd solution, save it, use MNE to average with old one
     fwd_copy['sol']['data'] *= 0.5
-    fname_copy = op.join(temp_dir, 'copy-fwd.fif')
+    fname_copy = str(tmpdir.join('copy-fwd.fif'))
```
Comment (agramfort) [inline mne/forward/tests/test_forward.py]: why do you still need to do this? it's only for py35?

**Comment by agramfort** [comment]
thx @larsoner

**Pull request merged**